File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
        if(head->next==nullptr) return head;
        ListNode* low = head;
        ListNode* high = head->next;
        while(high){
            ListNode* mid = new ListNode(__gcd(high->val , low->val));
            mid->next = high;
            low->next = mid;
            low = high;
            high = high->next;
        }
        return head;
    }
};

File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int f(int a,int b){
        while(b!=0){
            if(a>b)a=a-b;
            else b=b-a;
        }
        return a;
    }
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if(head==null || head.next==null)return head;
        ListNode dummy =new ListNode();
        ListNode curr =dummy;

        curr.next =head;
        curr =curr.next;

        head=head.next;

        while(head!=null){
            int gcd =f(curr.val,head.val);
            ListNode _gcd =new ListNode(gcd);
            
            curr.next =_gcd;
            curr=curr.next;

            curr.next=head;
            curr=curr.next;

            head=head.next;
        }
        return dummy.next;
    }
}